server/quota_test.go

```diff
@@ -0,0 +1,92 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"testing"
+
+	"github.com/google/trillian/quota"
+)
+
+func TestQuotaProviderRegistration(t *testing.T) {
+	for _, test := range []struct {
+		desc    string
+		reg     bool
+		wantErr bool
+	}{
+		{
+			desc:    "works",
+			reg:     true,
+			wantErr: false,
+		},
+		{
+			desc:    "unknown provider",
+			reg:     false,
+			wantErr: true,
+		},
+	} {
+		t.Run(test.desc, func(t *testing.T) {
+
+			called := false
+			name := test.desc
+
+			if test.reg {
+				RegisterQuotaManager(name, func() (quota.Manager, error) {
+					called = true
+					return nil, nil
+				})
+			}
+
+			_, err := NewQuotaManager(name)
+			if err != nil && !test.wantErr {
+				t.Fatalf("NewQuotaManager = %v, want no error", err)
+			}
+			if err == nil && test.wantErr {
+				t.Fatalf("NewQuotaManager = no error, want error")
+			}
+
+			if !called && !test.wantErr {
+				t.Fatal("Registered quota provider was not called")
+			}
+		})
+	}
+}
+
+func TestQuotaSystems(t *testing.T) {
+	RegisterQuotaManager("a", func() (quota.Manager, error) { return nil, nil })
+	RegisterQuotaManager("b", func() (quota.Manager, error) { return nil, nil })
+	qs := QuotaSystems()
+
+	if got, want := len(qs), 2; got < want {
```

Comment:
This test doesn't handle getting multiple copies of the same provider. E.g it would pass if you got 1000 copies of 'a' and 'b'.